test/e2e/apimachinery/custom_resource_definition.go

```diff
@@ -67,5 +77,151 @@ var _ = SIGDescribe("CustomResourceDefinition resources", func() {
 				}
 			}()
 		})
+
+		It("has OpenAPI spec served with CRD Validation chema", func() {
+			framework.SkipUnlessServerVersionGTE(crdPublishOpenAPIVersion, f.ClientSet.Discovery())
+
+			testcrd, err := framework.CreateTestCRD(f)
+			if err != nil {
+				return
+			}
+
+			// Test CRD don't have ValidationSchema setup. Here we hardcode a
+			// simple Schema for kubectl behavior testing.
+			// testcrd.Crd.Spec.Validation = fooSchema()
+			// patch := fmt.Sprintf("{\"spec\":{%s}}", utilyaml.ToJSON(validationSchema))
+			patch, err := utilyaml.ToJSON(validationSchema)
+			testcrd.Crd, err = testcrd.ApiExtensionClient.ApiextensionsV1beta1().
+				CustomResourceDefinitions().Patch(testcrd.GetMetaName(), types.MergePatchType, patch)
+			if err != nil {
+				framework.Failf("failed to patch CustomResourceDefinition: %v", err)
+			}
+			definitionKey := fmt.Sprintf("%s.%s.%s", testcrd.ApiGroup, testcrd.Versions[0].Name, testcrd.Kind)
+
+			// We use a wait.Poll block here because the kube-aggregator openapi
+			// controller takes time to rotate the queue and resync apiextensions-apiserver's spec
+			By("Waiting for test CRD's Schema to show up in OpenAPI spec")
+			lastMsg := ""
+			if err := wait.Poll(5*time.Second, 120*time.Second, func() (bool, error) {
+				bs, err := f.ClientSet.CoreV1().RESTClient().Get().AbsPath("openapi", "v2").DoRaw()
+				if err != nil {
+					return false, err
+				}
+				spec := spec.Swagger{}
+				if err := json.Unmarshal(bs, &spec); err != nil {
+					return false, err
+				}
+				if spec.SwaggerProps.Paths == nil {
+					lastMsg = "spec.SwaggerProps.Paths is nil"
+					return false, nil
+				}
+				d, ok := spec.SwaggerProps.Definitions[definitionKey]
+				if !ok {
+					lastMsg = fmt.Sprintf("spec.SwaggerProps.Definitions[\"%s\"] not found", definitionKey)
+					return false, nil
+				}
+				_, ok = d.Properties["status"]
+				if !ok {
+					lastMsg = fmt.Sprintf("spec.SwaggerProps.Definitions[\"%s\"].Properties[\"status\"] not found", definitionKey)
+					return false, nil
+				}
+				return true, nil
+			}); err != nil {
+				framework.Failf("failed to wait for apiserver to serve openapi spec for registered CRD: %v; lastMsg: %s", err, lastMsg)
+			}
+
+			By("Having kubectl understand the schema")
+			if _, err := framework.RunKubectl("explain", testcrd.GetPluralName()); err != nil {
+				framework.Failf("failed to explain CRD: %v", err)
+			}
+			if _, err := framework.RunKubectl("explain", testcrd.GetPluralName()+".spec"); err != nil {
```

Comment:
I would like this to actually check that a string defined in the spec shows up in the output - probably the field description and object description (the two most important fields)

Reply:
Updated

Reply:
@smarterclayton bump.  I really need this feature.